File: samples/javascript_nodejs/81.skills-skilldialog/dialogSkillBot/dialogs/activityRouterDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityTypes, InputHints, MessageFactory } = require('botbuilder');
const { ComponentDialog, DialogTurnStatus, WaterfallDialog } = require('botbuilder-dialogs');
const { LuisRecognizer } = require('botbuilder-ai');
const { BookingDialog } = require('./bookingDialog');

const ACTIVITY_ROUTER_DIALOG = 'activityRouterDialog';
const WATERFALL_DIALOG = 'waterfallDialog';
const BOOKING_DIALOG = 'bookingDialog';

/**
 * A root dialog that can route activities sent to the skill to different sub-dialogs.
 */
class ActivityRouterDialog extends ComponentDialog {
    constructor(conversationState, luisRecognizer = undefined) {
        super(ACTIVITY_ROUTER_DIALOG);

        if (!conversationState) throw new Error('[MainDialog]: Missing parameter \'conversationState\' is required');

        this.luisRecognizer = luisRecognizer;

        // Define the main dialog and its related components.
        // This is a sample "book a flight" dialog.
        this.addDialog(new BookingDialog())
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.processActivity.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async processActivity(stepContext) {
        // A skill can send trace activities, if needed.
        const traceActivity = {
            type: ActivityTypes.Trace,
            timestamp: new Date(),
            text: 'ActivityRouterDialog.processActivity()',
            label: `Got activityType: ${ stepContext.context.activity.type }`
        };
        await stepContext.context.sendActivity(traceActivity);

        switch (stepContext.context.activity.type) {
            case ActivityTypes.Event:
                return await this.onEventActivity(stepContext);
            case ActivityTypes.Message:
                return await this.onMessageActivity(stepContext);
            default:
                // We didn't get an activity type we can handle.
                await stepContext.context.sendActivity(
                    MessageFactory.text(
                        `Unrecognized ActivityType: "${ stepContext.context.activity.type }".`,
                        undefined,
                        InputHints.IgnoringInput
                    ));
                return { status: DialogTurnStatus.complete };
        }
    }

    /**
     * This method performs different tasks based on event name.
     */
    async onEventActivity(stepContext) {
        const activity = stepContext.context.activity;
        const traceActivity = {
            type: ActivityTypes.Trace,
            timestamp: new Date(),
            text: 'ActivityRouterDialog.onEventActivity()',
            label: `Name: ${ activity.name }, Value: ${ JSON.stringify(activity.value) }`
        };
        await stepContext.context.sendActivity(traceActivity);

        // Resolve what to execute based on the event name.
        switch (activity.name) {
            case 'BookFlight':
                return await this.beginBookFlight(stepContext);
            case 'GetWeather':
                return await this.beginGetWeather(stepContext);
            default:
                // We didn't get an event name we can handle.
                await stepContext.context.sendActivity(
                    MessageFactory.text(
                        `Unrecognized EventName: "${ stepContext.context.activity.name }".`,
                        undefined,
                        InputHints.IgnoringInput
                    ));
                return { status: DialogTurnStatus.complete };
        }
    }

    /**
     * This method just gets a message activity and runs it through LUIS.
     */
    async onMessageActivity(stepContext) {
        const activity = stepContext.context.activity;
        const traceActivity = {
            type: ActivityTypes.Trace,
            timestamp: new Date(),
            text: 'ActivityRouterDialog.onMessageActivity()',
            label: `Text: ${ activity.text }, Value: ${ JSON.stringify(activity.value) }`
        };
        await stepContext.context.sendActivity(traceActivity);

        if (!this.luisRecognizer || !this.luisRecognizer.isConfigured) {
            await stepContext.context.sendActivity(MessageFactory.text(
                'NOTE: LUIS is not configured. To enable all capabilities, please add \'LuisAppId\', \'LuisAPIKey\' and \'LuisAPIHostName\' to the appsettings.json file.',
                undefined,
                InputHints.IgnoringInput
            ));
        } else {
            // Call LUIS with the utterance.
            const luisResult = await this.luisRecognizer.executeLuisQuery(stepContext.context);
            const topIntent = LuisRecognizer.topIntent(luisResult);

            // Create a message showing the LUIS result.
            let resultString = '';
            resultString += `LUIS results for "${ activity.text }":\n`;
            resultString += `Intent: "${ topIntent }", Score: ${ luisResult.intents[topIntent].score }\n`;

            await stepContext.context.sendActivity(MessageFactory.text(resultString, undefined, InputHints.IgnoringInput));

            switch (topIntent.intent) {
                case 'BookFlight':
                    return await this.beginBookFlight(stepContext);
                case 'GetWeather':
                    return await this.beginGetWeather(stepContext);
                default:
                    // Catch all for unhandled intents.
                    const didntUnderstandMessageText = `Sorry, I didn't get that. Please try asking in a different way (intent was ${ topIntent.intent })`;
                    const didntUnderstandMessage = MessageFactory.text(didntUnderstandMessageText, didntUnderstandMessageText, InputHints.IgnoringInput);
                    await stepContext.context.sendActivity(didntUnderstandMessage);
                    break;
            }
        }

        return { status: DialogTurnStatus.complete };
    }

    async beginGetWeather(stepContext) {
        const activity = stepContext.context.activity;
        const location = activity.value || {};

        // We haven't implemented the GetWeatherDialog so we just display a TODO message.
        const getWeatherMessageText = `TODO: get weather for here (lat: ${ location.latitude }, long: ${ location.longitude })`;
        const getWeatherMessage = MessageFactory.text(getWeatherMessageText, getWeatherMessageText, InputHints.IgnoringInput);
        await stepContext.context.sendActivity(getWeatherMessage);
        return { status: DialogTurnStatus.complete };
    }

    async beginBookFlight(stepContext) {
        const activity = stepContext.context.activity;
        const bookingDetails = activity.value || {};

        // Start the booking dialog.
        const bookingDialog = this.findDialog(BOOKING_DIALOG);
        return await stepContext.beginDialog(bookingDialog.id, bookingDetails);
    }
}

module.exports.ActivityRouterDialog = ActivityRouterDialog;

File: samples/java_springboot/81.skills-skilldialog/dialog-skill-bot/src/main/java/com/microsoft/bot/sample/dialogskillbot/dialogs/ActivityRouterDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MT License.

package com.microsoft.bot.sample.dialogskillbot.dialogs;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.builder.RecognizerResult;
import com.microsoft.bot.builder.TurnContext;
import com.microsoft.bot.dialogs.ComponentDialog;
import com.microsoft.bot.dialogs.Dialog;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.dialogs.DialogTurnStatus;
import com.microsoft.bot.dialogs.WaterfallDialog;
import com.microsoft.bot.dialogs.WaterfallStep;
import com.microsoft.bot.dialogs.WaterfallStepContext;
import com.microsoft.bot.restclient.serializer.JacksonAdapter;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.ActivityTypes;
import com.microsoft.bot.schema.InputHints;
import com.microsoft.bot.schema.Serialization;

/**
 * A root dialog that can route activities sent to the skill to different
 * sub-dialogs.
 */
public class ActivityRouterDialog extends ComponentDialog {

    private final DialogSkillBotRecognizer luisRecognizer;

    public ActivityRouterDialog(DialogSkillBotRecognizer luisRecognizer) {
        super("ActivityRouterDialog");
        this.luisRecognizer = luisRecognizer;

        addDialog(new BookingDialog());
        List<WaterfallStep> stepList = new ArrayList<WaterfallStep>();
        stepList.add(this::processActivity);
        addDialog(new WaterfallDialog("WaterfallDialog", stepList));

        // The initial child Dialog to run.
        setInitialDialogId("WaterfallDialog");
    }

    private CompletableFuture<DialogTurnResult> processActivity(WaterfallStepContext stepContext) {
        // A skill can send trace activities, if needed.
        TurnContext.traceActivity(
            stepContext.getContext(),
            String.format(
                "{%s}.processActivity() Got ActivityType: %s",
                this.getClass().getName(),
                stepContext.getContext().getActivity().getType()
            )
        );

        switch (stepContext.getContext().getActivity().getType()) {
            case ActivityTypes.EVENT:
                return onEventActivity(stepContext);

            case ActivityTypes.MESSAGE:
                return onMessageActivity(stepContext);

            default:
                String defaultMessage = String
                    .format("Unrecognized ActivityType: \"%s\".", stepContext.getContext().getActivity().getType());
                // We didn't get an activity type we can handle.
                return stepContext.getContext()
                    .sendActivity(MessageFactory.text(defaultMessage, defaultMessage, InputHints.IGNORING_INPUT))
                    .thenCompose(result -> {
                        return CompletableFuture.completedFuture(new DialogTurnResult(DialogTurnStatus.COMPLETE));
                    });

        }
    }

    // This method performs different tasks super. on the event name.
    private CompletableFuture<DialogTurnResult> onEventActivity(WaterfallStepContext stepContext) {
        Activity activity = stepContext.getContext().getActivity();
        TurnContext.traceActivity(
            stepContext.getContext(),
            String.format(
                "%s.onEventActivity(), label: %s, Value: %s",
                this.getClass().getName(),
                activity.getName(),
                GetObjectAsJsonString(activity.getValue())
            )
        );

        // Resolve what to execute super. on the event name.
        switch (activity.getName()) {
            case "BookFlight":
                return beginBookFlight(stepContext);

            case "GetWeather":
                return beginGetWeather(stepContext);

            default:
                String message = String.format("Unrecognized EventName: \"%s\".", activity.getName());
                // We didn't get an event name we can handle.
                stepContext.getContext().sendActivity(MessageFactory.text(message, message, InputHints.IGNORING_INPUT));
                return CompletableFuture.completedFuture(new DialogTurnResult(DialogTurnStatus.COMPLETE));
        }
    }

    // This method just gets a message activity and runs it through LUS.
    private CompletableFuture<DialogTurnResult> onMessageActivity(WaterfallStepContext stepContext) {
        Activity activity = stepContext.getContext().getActivity();
        TurnContext.traceActivity(
            stepContext.getContext(),
            String.format(
                "%s.onMessageActivity(), label: %s, Value: %s",
                this.getClass().getName(),
                activity.getName(),
                GetObjectAsJsonString(activity.getValue())
            )
        );

        if (!luisRecognizer.getIsConfigured()) {
            String message = "NOTE: LUIS instanceof not configured. To enable all capabilities, add 'LuisAppId',"
                + " 'LuisAPKey' and 'LuisAPHostName' to the appsettings.json file.";
            return stepContext.getContext()
                .sendActivity(MessageFactory.text(message, message, InputHints.IGNORING_INPUT))
                .thenCompose(
                    result -> CompletableFuture.completedFuture(new DialogTurnResult(DialogTurnStatus.COMPLETE))
                );
        } else {
            // Call LUIS with the utterance.
            return luisRecognizer.recognize(stepContext.getContext(), RecognizerResult.class)
                .thenCompose(luisResult -> {
                    // Create a message showing the LUS results.
                    StringBuilder sb = new StringBuilder();
                    sb.append(String.format("LUIS results for \"%s\":", activity.getText()));

                    sb.append(
                        String.format(
                            "Intent: \"%s\" Score: %s",
                            luisResult.getTopScoringIntent().intent,
                            luisResult.getTopScoringIntent().score
                        )
                    );

                    return stepContext.getContext()
                        .sendActivity(MessageFactory.text(sb.toString(), sb.toString(), InputHints.IGNORING_INPUT))
                        .thenCompose(result -> {
                            switch (luisResult.getTopScoringIntent().intent.toLowerCase()) {
                                case "bookflight":
                                    return beginBookFlight(stepContext);

                                case "getweather":
                                    return beginGetWeather(stepContext);

                                default:
                                    // Catch all for unhandled intents.
                                    String didntUnderstandMessageText = String.format(
                                        "Sorry, I didn't get that. Please try asking in a different "
                                        + "way (intent was %s)",
                                        luisResult.getTopScoringIntent().intent
                                    );
                                    Activity didntUnderstandMessage = MessageFactory.text(
                                        didntUnderstandMessageText,
                                        didntUnderstandMessageText,
                                        InputHints.IGNORING_INPUT
                                    );
                                    return stepContext.getContext()
                                        .sendActivity(didntUnderstandMessage)
                                        .thenCompose(
                                            stepResult -> CompletableFuture
                                                .completedFuture(new DialogTurnResult(DialogTurnStatus.COMPLETE))
                                        );

                            }
                        });
                    // Start a dialog if we recognize the intent.
                });
        }
    }

    private static CompletableFuture<DialogTurnResult> beginGetWeather(WaterfallStepContext stepContext) {
        Activity activity = stepContext.getContext().getActivity();
        Location location = new Location();
        if (activity.getValue() != null) {
            try {
                location = Serialization.safeGetAs(activity.getValue(), Location.class);
            } catch (JsonProcessingException e) {
                // something went wrong, so we create an empty Location so we won't get a null
                // reference below when we acess location.
                location = new Location();
            }

        }

        // We haven't implemented the GetWeatherDialog so we just display a TODO
        // message.
        String getWeatherMessageText = String
            .format("TODO: get weather for here (lat: %s, long: %s)", location.getLatitude(), location.getLongitude());
        Activity getWeatherMessage =
            MessageFactory.text(getWeatherMessageText, getWeatherMessageText, InputHints.IGNORING_INPUT);
        return stepContext.getContext().sendActivity(getWeatherMessage).thenCompose(result -> {
            return CompletableFuture.completedFuture(new DialogTurnResult(DialogTurnStatus.COMPLETE));
        });

    }

    private CompletableFuture<DialogTurnResult> beginBookFlight(WaterfallStepContext stepContext) {
        Activity activity = stepContext.getContext().getActivity();
        BookingDetails bookingDetails = new BookingDetails();
        if (activity.getValue() != null) {
            try {
                bookingDetails = Serialization.safeGetAs(activity.getValue(), BookingDetails.class);
            } catch (JsonProcessingException e) {
                // we already initialized bookingDetails above, so the flow will run as if
                // no details were sent.
            }
        }

        // Start the booking dialog.
        Dialog bookingDialog = findDialog("BookingDialog");
        return stepContext.beginDialog(bookingDialog.getId(), bookingDetails);
    }

    private String GetObjectAsJsonString(Object value) {
        try {
            return new JacksonAdapter().serialize(value);
        } catch (IOException e) {
            return null;
        }
    }
}
